PFW-1536: Reprint function implemented for MK3S/MK3S+

jfestrada opened the pull request:
This PR is a the same than #3071 but using the current repository of my fork of your firmware

To support this functionality I have added a **extern bool variable** which can be read/set from cmdqueue.cpp and **ultralcd.cpp** so when the card reader reaches the EOF the variable isPrintFinished will be set to true and the main menu will show the **REPRINT** option.
How do I know which file we need to reprint? All the information needed to perform the reprint was stored on the EEPROM in case of a power panic and the fw recover this information in order to start the print from SD again, so using that information the function reprint_from_eeprom() has been implemented.

Translations:
- [X] :czech_republic: [same as buddy firmware](https://github.com/prusa3d/Prusa-Firmware-Buddy/blob/bfb0ffc745ee6546e7efdba618d0e7c0f4c909cd/src/lang/po/cs/Prusa-Firmware-Buddy_cs.po#L3647-L3648)
- [X] :de: [same as buddy firmware](https://github.com/prusa3d/Prusa-Firmware-Buddy/blob/bfb0ffc745ee6546e7efdba618d0e7c0f4c909cd/src/lang/po/de/Prusa-Firmware-Buddy_de.po#L3737-L3738)
- [X] :es: [same as buddy firmware](https://github.com/prusa3d/Prusa-Firmware-Buddy/blob/bfb0ffc745ee6546e7efdba618d0e7c0f4c909cd/src/lang/po/es/Prusa-Firmware-Buddy_es.po#L3710-L3711)
- [X] :fr: [same as buddy firmware](https://github.com/prusa3d/Prusa-Firmware-Buddy/blob/bfb0ffc745ee6546e7efdba618d0e7c0f4c909cd/src/lang/po/fr/Prusa-Firmware-Buddy_fr.po#L3676-L3677)
- [X] :it: [same as buddy firmware](https://github.com/prusa3d/Prusa-Firmware-Buddy/blob/bfb0ffc745ee6546e7efdba618d0e7c0f4c909cd/src/lang/po/it/Prusa-Firmware-Buddy_it.po#L3698-L3699)
- [X] :poland: [same as buddy firmware](https://github.com/prusa3d/Prusa-Firmware-Buddy/blob/bfb0ffc745ee6546e7efdba618d0e7c0f4c909cd/src/lang/po/pl/Prusa-Firmware-Buddy_pl.po#L3679-L3680)
- [X] :netherlands: by @3d-gussner @vintagepc please check [Firmware_nl.po](https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-92de134d7cfe75b62e86a4e83716cd95550b11f2ddee228150a9c6faa4d6a34bR2585-R2586)
- [ ] :croatia: @Prime1910 please check [Firmware_hr.po](https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-487d5469566575627c7dc47abf437c52e890de941c1fd8d4db5a2a4d6ef08d64R2577-R2578)
- [x] :hungary: @AttilaSVK please check [Firmware_hu.po](https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-3a7d51d07fb16c60d273b7986aa88a7892939b19a9bad4f3501a153ffb0604e9R2582-R2583)
- [ ] :norway: @OS-kar and @trondkla please check [Firmware_no.po](https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-22e5dd5d54111e837f3b40db65f2a126b09cc4157dcfa1f34b6e5dde0bf4a607R2559-R2560)
- [x] :romania: @leptun and @Hauzman please check [Firmware_ro.po](https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-6774f424618c20ccb1d99bf93cb34bab0b6ee03a55868fa7021a11ef6c28667fR2583-R2584)
- [x] :slovakia: @ingbrzy and @shatter136 please check [Firmware_sk.po](https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-9584ba5c6a2b7618c16dc96fa21d692e9d5d69c393c55f37166bb05e98557fbcR2565-R2566)
- [ ] :sweden: @Painkiller56 please check [Firmware_sv.po](https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-cf46655e0192f28112da2985d28dc728949010730a0ac629dbb7049f9f53a357R2572-R2573)

gudnimg committed:
optimisation: use disable_heater() in more places

Change in memory:
Flash: -84 bytes
SRAM: 0 bytes

3d-gussner committed:
Add M125

3d-gussner committed:
Fix float
Thanks to @gudnimg for pointing out

gudnimg committed:
Fix filament change behavior when print is paused

Reimplementing PR 2390
https://github.com/prusa3d/Prusa-Firmware/pull/2390
from @wavexx

gudnimg committed:
power panic: only send M23 for SD prints

If the saved printing type was USB, then EEPROM_FILENAME
does not contain anything. The firmware should also
not be trying to open a file on a SD card which is maybe
not even mounted.

Change in memory:
Flash: +12 bytes
SRAM: 0 bytes

gudnimg committed:
power panic: read EEPROM_FILENAME until \0 is seen

Also removed debug logs

gudnimg committed:
Refactor menu_action_sdfile

gudnimg committed:
power panic: simplify recovering SD filename

We can just read the whole EEPROM block since short filenames
are always null terminated. strcat_P will then apply the file extension
at the correct position.

Change in memory:
Flash: -24 bytes
SRAM: 0 bytes

gudnimg committed:
power panic: read whole directory name block from EEPROM

There is no need to read one byte at a time. We can simply
read the whole block in one go. This saves some flash memory.

Change in memory:
Flash: -18 bytes
SRAM: 0 bytes

gudnimg committed:
mmu: Add CheckErrorScreenUserInput

Due to differences in 8-bit FW I had to implement two new get functions

* GetCommandInProgress
* GetLastErrorCode

gudnimg committed:
mmu: formatting

gudnimg committed:
mmu: add DisableMMUInSettings

power_on should not be modifying
EEPROM_MMU_ENABLED. The code is never
executed unless it's already been set.

Only disable EEPROM_MMU_ENABLED through Buttons::DisableMMU

Change in memory:
Flash: -18 bytes
SRAM: 0 bytes

gudnimg committed:
mmu: add ResetCommunicationTimeoutAttempts

Adding __attribute__((noinline)) saves 14 bytes of flash

Change in memory:
Flash: +34 bytes
SRAM: 0 bytes

gudnimg committed:
mmu: remove ResetOnExit

No change in memory

gudnimg committed:
mmu: remove obsolete EEPROM comments

gudnimg committed:
mmu: remove isErrorScreenRunning()

No change in memory

3d-gussner committed:
Basic Printer name in EEPROM
- Output with M115 works
- ToDo
  - LCD screen
  - M550 gcode
  - Detect length of EEPROM CUSTOM_MENDLE_NAME

gudnimg committed:
PFW-1523 Implement M79 timer for PrusaLink

The idea is to have the host ping the printer periodically with a M79 to
enable certain features/UI. Using the usb_timer is not a good solution
for this as it depends on seeing a 'G' character

The LCD code, or whatever code is implementing the new functionality
will need to include host.h and check whether M79_timer_get_status()
returns 0 (timer not running) or 1 (timer is running).

I created a new file for the code host.cpp which we can use to expand
host related functionality and not clutter Marlin_main.cpp further.

Change in memory:
Flash: +104 bytes
SRAM: +5 bytes

gudnimg committed:
optimisation: simplify two LCD messages

There is no need to restrict how often the message is rendered.
It was being restricted to render every 1 second. We don't do
this for most of the static menus. So I propose this 1 second period
be removed.

Tested on MK3S+

Change in memory:
Flash: -168 bytes
SRAM: -8 bytes

jfestrada committed:
Reprint function integrated on 3.13v

jfestrada committed:
Fix the length of the PL translation of MSG_REPRINT

jfestrada committed:
Fix SV translation length of MSG_REPRINT

jfestrada commented:
@3d-gussner could you share an example of the `// action:reprint`?

gudnimg commented on Firmware/ultralcd.cpp:
```diff
@@ -7427,3 +7437,52 @@ void lcd_heat_bed_on_load_toggle()
+
+	//cmdqueue_reset();
+
+	depth = eeprom_read_byte((uint8_t*)EEPROM_DIR_DEPTH);
+
+	MYSERIAL.println(int(depth));
+	for (int i = 0; i < depth; i++) {
+		for (int j = 0; j < 8; j++) {
+			dir_name[j] = eeprom_read_byte((uint8_t*)EEPROM_DIRS + j + 8 * i);
+		}
+		dir_name[8] = '\0';
+		MYSERIAL.println(dir_name);
+		// strcpy(dir_names[i], dir_name);
+		card.chdir(dir_name, false);
+	}
+
+	for (int i = 0; i < 8; i++) {
+		filename[i] = eeprom_read_byte((uint8_t*)EEPROM_FILENAME + i);
+	}
+	filename[8] = '\0';
```
We can probably save quite a bit of memory by re-using `restore_file_from_sd()` here from one of my PRs: https://github.com/prusa3d/Prusa-Firmware/pull/4437/files

jfestrada replied:
@gudnimg the only lack about your approach is related with the files which have the _.g_ extension instead of _.gco_
That's something that were already discussed on the original PR #3071, here is the [link for the comments](https://github.com/prusa3d/Prusa-Firmware/pull/3071#discussion_r615726244)
Because of that I added this block I know that it was not the more elegant solution, but it works

	strcpy(altfilename,filename);
	if (!card.FileExists(altfilename))
	{
		strcat_P(filename, PSTR(".gco"));
		if (card.FileExists(filename))
		{
			strcpy(altfilename,filename);
		}else
		{
			strcat_P(altfilename, PSTR(".g"));
		}
	}

gudnimg replied:
I see 🤔 The power panic code should also handle the `.g` extension but it always assume it is `.gco` (Its been that way for years). I think it is time we add the short file name extension into EEPROM. 

We can’t expand `EEPROM_FILENAME` since it will affect adjacent memory. But we can add the extension in a different location. I can fix it in my PR since your PR is unrelated to power panic functionality. 

I think we can reduce the flash increase of your PR from +456 bytes to something closer to +200 bytes. 

We can probably merge your PR first and then later I can create a separate PR to “optimise” it because the flash memory we have left is precious.

jfestrada replied:
That make sense, if you need anything from my side just tell me.

gudnimg replied:
Resolving this thread for now.

gudnimg replied:
I pushed a fix to my PR to support any valid DOS 8.3 file extension. Only tested `".gco"` and `".g"` on my end and it works. I can confirm the current MK3 branch does not work for `".g"`

This did cost ~100 bytes of flash though, but I think it will be worth it since the Re-print functionality can re-use the code.

gudnimg commented on Firmware/ultralcd.cpp:
```diff
@@ -7427,3 +7437,52 @@ void lcd_heat_bed_on_load_toggle()
+		}
+	}
+	MYSERIAL.print(altfilename);
+	sprintf_P(cmd, PSTR("M23 %s"), altfilename);
+	enquecommand(cmd);
+  	sprintf_P(cmd, PSTR("M24"));
+	enquecommand(cmd);
```
Please use `enquecommandf_P` and remove `cmd` buffer. This should save some flash memory.
```c
// M23: Select SD file
enquecommandf_P(MSG_M23, altfilename);
// M24: Start/resume SD print
enquecommand_P(MSG_M24);
```

jfestrada replied:
I made that change

gudnimg replied:
Nice that’s 56 bytes of flash down ❤️

jfestrada committed:
Change to enquecommandf_P and enquecommand_P

gudnimg committed:
Implement Marlin's print job timer

I changed the code a bit so it would compile
in Prusa Firmware.

The paused duration is no longer included in the
print time statistics.

Change in memory:
Flash: -122 bytes
SRAM: 0 bytes

gudnimg commented on Firmware/ultralcd.cpp:
```diff
@@ -7427,3 +7437,52 @@ void lcd_heat_bed_on_load_toggle()
+			dir_name[j] = eeprom_read_byte((uint8_t*)EEPROM_DIRS + j + 8 * i);
+		}
+		dir_name[8] = '\0';
+		MYSERIAL.println(dir_name);
```
Minor nitpick. But I feel all the calls to `MYSERIAL` in this function should be removed. They increase flash memory usage unnecessarily.

In my opinion they are local debugging code.

```c
MYSERIAL.println(int(depth));
MYSERIAL.println(dir_name);
MYSERIAL.print(altfilename);
```

jfestrada replied:
It was for debug purpose, already deleted

gudnimg replied:
Sorry for picking at these small things but there are two `MYSERIAL` lines left.

jfestrada replied:
> Sorry for picking at these small things but there are two `MYSERIAL` lines left.

I'm sorry, I have removed already.

gudnimg replied:
Thanks :) Resolving this thread. 

I have yet to test the PR on hardware, will approve later this week if all works as expected.

jfestrada replied:
Maybe would be nice to check the translations of "REPRINT" because I used google translator

gudnimg commented on Firmware/ultralcd.cpp:
```diff
@@ -7427,3 +7437,52 @@ void lcd_heat_bed_on_load_toggle()
+
+	enableReprint=false;
+
+	//cmdqueue_reset();
```
Please remove commented code :)

jfestrada replied:
I have removed the commented code

gudnimg committed:
power panic: support any valid DOS 8.3 extension

When a SD file is selected to print
save the DOS 8.3 extension into EEPROM.

After a power outage, the correct file extension is then
selected instead of always assuming it's ".gco"

This allows users to recover ".g" files.

Change in memory:
Flash: +104 bytes
SRAM: 0 bytes

jfestrada committed:
Remove MYSERIAL.println code (debug purpose) and old commented code

jfestrada committed:
Remove some MYSERIAL calls

3d-gussner commented on Firmware/messages.cpp:
```diff
@@ -187,7 +187,7 @@ extern const char MSG_LOAD_ALL[] PROGMEM_I1 = ISTR("Load All"); ////MSG_LOAD_ALL
 extern const char MSG_NOZZLE_CNG_READ_HELP [] PROGMEM_I1 = ISTR("For a Nozzle change please read\nprusa.io/nozzle-mk3s");////MSG_NOZZLE_CNG_READ_HELP c=20 r=4
 extern const char MSG_NOZZLE_CNG_CHANGED [] PROGMEM_I1 = ISTR("Hotend at 280C! Nozzle changed and tightened to specs?");////MSG_NOZZLE_CNG_CHANGED c=20 r=6
-
+extern const char MSG_REPRINT [] PROGMEM_I1 = ISTR("Reprint"); ////MSG_REPRINT c=7
```
As this is a submenu the allowed char length is `c=18`

jfestrada replied:
Ok, I suppose that I had to specify the longer length, I will do the change.

3d-gussner replied:
The pot file defines all messages and po files are based on it.
That is why we have the two update-po/t.py scripts to keep them correct.
Editing the po/t files manually can break things badly.

gudnimg replied:
Reminder, that `c=7` is still present, should be `c=18`

3d-gussner commented on lang/po/Firmware.pot:
```diff
@@ -2460,3 +2460,7 @@ msgstr ""
 msgid "MK3 firmware detected on MK3S printer"
 msgstr ""
+
+#. MSG_REPRINT c=7 
```
Submenu can be `c=18` long

3d-gussner replied:
`.pot` files and `.po`files have to be updated with `lang/update-pot.py` and `lang/update-po-py`.

3d-gussner reviewed (changes requested):
The language `Firmware.pot` and `Firmware_xx.po` have to be update with the scripts `update-pot.py` and `update.po.py`.
I will change it for you if that is okay.

Submenu  can be 18 character long and NEED to have same `c=` values in the `.po` files as it is in `.pot` file.

3d-gussner commented:
The SD card print from the LCD screen is saved to eeprom, but what about the SD card prints started from a host using `M23` and `M24` these aren't covered, or are they?

3d-gussner commented:
After a USB/Host print (which is streaming the gcode over serial line) the `Reprint` menu should send a serial message `// action:reprint` to the host. Then the host has to do/run its own sequence depending on this `action`.

We can use the existing  `M118 A1 action:reprint` to send the message `// action:reprint` via serial line.

So the Reprint menu on LCD should:
- restart the SD print
  - When the previous print was a SD print that has been selected via LCD menu before :heavy_check_mark: 
    - We save already the information in eeprom and this PR uses this. That enables the option to reprint also after reset or power cycle
      - I think in this case it would be useful to show which file will be reprinted with an yes/no option to continue with `yes` if times out 
  - When the previous print was a SD print that has been started via `M23` and `M24` gcode commands from USB/host system before :construction: 
     - We don't save this information to eeprom, and not sure if we should do that as well like selecting from LCD screen.
       - We could read from the memory only.  
       - If not implemented the reprint after a reset or power cycle would start the last LCD selected print and not the last via gcode started SD print.
- When the previous print was an USB/Host streamed print send `// action:reprint`. :construction:

jfestrada commented:
> The SD card print from the LCD screen is saved to eeprom, but what about the SD card prints started from a host using `M23` and `M24` these aren't covered, or are they?

Only if the print is launched from the remote host but with a local SD file, in other case I doubt it should be covered, I mean, when you are printing from SD the main user input is using the LCD, there is where the Reprint function make sense because the user does not need to search the file again on the SD using the LCD menu. When the user is printing from a remote host, their main input is made through that host, so make sense that should be that host which take the control of the Reprint action, usually in octoprint you just need to hit the print button again. Is similar than in mainsail or fluid interfaces and I suppose than in prusaconnect will be the same.
![image](https://github.com/prusa3d/Prusa-Firmware/assets/7290345/a5b2b841-0f98-448b-818e-ef62b7329171)
Anyway in the case that you are executing the printing from the SD but from a remote host (octoprint, prusaconnect, ...), that should show the reprint menu in the LCD because is similar to execute it from the printer directly, and the _enableReprint_ variable becomes to true when the cmdqueue became empty and the card report the EOF.

```
if(card.eof())
  {
      // file was fully buffered, but commands might still need to be planned!
      // do *not* clear sdprinting until all SD commands are consumed to ensure
      // SD state can be resumed from a saved printing state. sdprinting is only
      // cleared by printingHasFinished after peforming all remaining moves.
      if(!cmdqueue_calc_sd_length())
      {
          // queue is complete, but before we process EOF commands prevent
          // re-entry by disabling SD processing from any st_synchronize call
          card.closefile();
          enableReprint=true;
``` 

Also as the printer is using the SD to print the file the firmware will store the file and the line on eeprom.

3d-gussner commented:
@jfestrada The "Reprint" option on LCD is for the User after clearing the bed.

1. Remove print
2. Clean sheet
3. Press Reprint on LCD as they are already there
   a. without having to go back to the computer and press the Reprint button in PrusaLink/PrusaConnect/OctoPrint/other host system.
   b. without having to select the SD again in the dirs etc.

3d-gussner commented:
@jfestrada I personally prefer to print from SD and monitor with PrusaLinl/PrusaConnect or Octoprint. When anything happens to the host/network the print continues! And with the RPi port the printer doesn't restart as with USB (you can switch that off BTW) and can continue to monitor the print.

So if I had a farm:

- I would store all files on SD card
- Use PrusaLink/PrusaConnect or Octoprint as central monitoring system.
- Start these SD prints from HOST.

jfestrada commented:
> After a USB/Host print (which is streaming the gcode over serial line) the `Reprint` menu should send a serial message `// action:reprint` to the host. Then the host has to do/run its own sequence depending on this `action`.
> 
> We can use the existing `M118 A1 action:reprint` to send the message `// action:reprint` via serial line.

OK, I think that could be easy to implement and nice to have.

> So the Reprint menu on LCD should:
> 
>     * restart the SD print
>       
>       * When the previous print was a SD print that has been selected via LCD menu before ✔️
>         
>         * We save already the information in eeprom and this PR uses this. That enables the option to reprint also after reset or power cycle

As it is implemented currently, after a power cycle or eject of the SD the Reprint is disabled, why? because you can't be sure if the last print was OK or if the inserted SD is the same. Would be nice to have the reprint enabled after a power cycle? yes but the firmware will need to check that the last print was finished and that the file exists on the SD before to show the menu for the Reprint. Now the question is, should the reprint be enabled after a powercyle?

>           
>           * I think in this case it would be useful to show which file will be reprinted with an yes/no option to continue with `yes` if times out

Ok, I agree with that

>       * When the previous print was a SD print that has been started via `M23` and `M24` gcode commands from USB/host system before 🚧
>         
>         * We don't save this information to eeprom, and not sure if we should do that as well like selecting from LCD screen.
>           
>           * We could read from the memory only.
>           * If not implemented the reprint after a reset or power cycle would start the last LCD selected print and not the last via gcode started SD print.
currently that will not happen because after reset or powercycle the variable enableReprint is set to false.

Should the reprint be enabled after a powercyle?

>     * When the previous print was an USB/Host streamed print send `// action:reprint`. 🚧
Ok, I agree

jfestrada commented:
> @jfestrada I personally prefer to print from SD and monitor with PrusaLinl/PrusaConnect or Octoprint. When anything happens to the host/network the print continues! And with the RPi port the printer doesn't restart as with USB (you can switch that off BTW) and can continue to monitor the print.
> 
> So if I had a farm:
> 
>     * I would store all files on SD card
> 
>     * Use PrusaLink/PrusaConnect or Octoprint as central monitoring system.
> 
>     * Start these SD prints from HOST.

That's also my preference, so we have the same point of view

jfestrada commented:
So I will take care of:

* Update language files
* Send the `action:reprint` when the print has been streamed
* Implement the yes/no question with the name of the file to be reprinted in case of a SD print

My question now is, should be enabled the reprint after a power cycle?

3d-gussner commented:
> Implement the yes/no question with the name of the file to be reprinted in case of a SD print
> My question now is, should be enabled the reprint after a power cycle?

You made a point that the reprint after power cycle isn't really needed, and so no need to add the extra message.
This was only a though as we have it in eeprom to use it, but as the bool reprint is reset after power cycle i would say no need for additional code and usage of resources.
If users ask for that we can implement it later.

arekm commented:
> Anyway in the case that you are executing the printing from the SD but from a remote host (octoprint, prusaconnect, ...), that should show the reprint menu

Just to be clear on this - so prints started remotely (from octoprint for example) via M23/M24 are covered and "Reprint" will show up - correct?

jfestrada commented:
> > Anyway in the case that you are executing the printing from the SD but from a remote host (octoprint, prusaconnect, ...), that should show the reprint menu
> 
> Just to be clear on this - so prints started remotely (from octoprint for example) via M23/M24 are covered and "Reprint" will show up - correct?

Yes, those one are covered with the _Reprint_ function through the LCD menu

3d-gussner committed:
Set Ready

gudnimg committed:
MBL: optimise EEPROM read/writes

gudnimg committed:
TM: Optimise EEPROM read and writes

gudnimg committed:
MMU: Set PAUSED_FOR_USER state on MMU error screen

Setting this state notifies PrusaLink/PrusaConnect
that the printer is waiting for user input (attention).

Change in memory:
Flash: +12 bytes
SRAM: 0 bytes

gudnimg committed:
Fix regression in MMU enable menu

Fixes a regression introduced in https://github.com/prusa3d/Prusa-Firmware/pull/4454

3d-gussner commented:
@jfestrada Do you have any update commits for this PR?

jfestrada committed:
- Serial message --> /// action:reprint done
- Yes/no question --> Need to fix the bug related with the filename print --> if (lcd_show_fullscreen_message_yes_no_and_wait_P(altfilename, false, LCD_LEFT_BUTTON_CHOICE)==LCD_LEFT_BUTTON_CHOICE)

jfestrada commented:
> @jfestrada Do you have any update commits for this PR?

I was working on this yesterday (I have made the push just to share the code). Let me explain.

- I have changed the initialization of the _enableReprint_ variable on the cmdqueue because can't be done when the file is closed on the SD, now must be set to true when the buffer is empty, is there any other way to detect the end of a streamed gcode?
- I have added the reference to powerpanic.h on the ultralcd.cpp because I have to know which kind of print I'm doing in order to send the /// action:reprint or just start the reprint from the SD. 
- Now I was dealing with the yes/no question and the name of the file, is there an easy way to reuse the function lcd_show_fullscreen_message_yes_no_and_wait_P with a char[13] as first argument? I do not want to reinvent the wheel.  Is there is no an easy way I will use the function lcd_printf_P.

Also I have to test the action:reprint with octoprint.

3d-gussner commented:
@jfestrada Thanks for the update

> Now I was dealing with the yes/no question and the name of the file, is there an easy way to reuse the function lcd_show_fullscreen_message_yes_no_and_wait_P with a char[13] as first argument? I do not want to reinvent the wheel. Is there is no an easy way I will use the function lcd_printf_P.

Let us skip this request as we DON'T reprint from EEPROM after a reboot see above

> You made a point that the reprint after power cycle isn't really needed, and so no need to add the extra message.

3d-gussner commented:
> I have added the reference to powerpanic.h on the ultralcd.cpp because I have to know which kind of print I'm doing in order to send the /// action:reprint or just start the reprint from the SD.

We deal with the Host USB printing power panic recovery in a different way, so no need to send `// action:reprint` when power panic recovers.
Or do I miss a case where we should have the reprint option after a power panic?

3d-gussner committed:
Add en- and disable MMU to M709
Add X42 to erase the MMU eeprom

jfestrada committed:
Added support for reprint and streamed GCODE

jfestrada committed:
Execution of lang/update-pot.py and lang/update-po.py

jfestrada committed:
Fix execution of update-po.py

gudnimg committed:
PFW-1523 Refactor parsing quoted strings

My plan is to re-use this function in M79
in a later commit. The firmware doesn't
have a dedicated parser like Marlin 2.1
so this is my attempt to not duplicate the parsing of a quoted string in G-codes

Change in memory (MK3S+ Multilang):
Flash: -50 bytes
SRAM: 0 bytes

gudnimg committed:
PFW-1523 Implement S parameter for M79

Change in memory:
Flash: +112 bytes
SRAM: +3 bytes

gudnimg committed:
PFW-1523 Implement S parameter for M79

Change in memory:
Flash: +112 bytes
SRAM: +3 bytes

gudnimg committed:
PFW-1523 Force inline unquoted_string constructor

Saves 36 bytes of flash

3d-gussner committed:
Update requested changes

3d-gussner commented on Firmware/cardreader.h:
```diff
@@ -83,6 +83,9 @@ class CardReader
   bool ToshibaFlashAir_isEnabled() const { return card.getFlashAirCompatible(); }
   void ToshibaFlashAir_enable(bool enable) { card.setFlashAirCompatible(enable); }
   bool ToshibaFlashAir_GetIP(uint8_t *ip);
+  
```
Remove white space

3d-gussner commented on Firmware/ultralcd.cpp:
```diff
@@ -5181,6 +5185,17 @@ static void lcd_main_menu()
+    if(!printer_active() && enableReprint && card.cardOK)
+    {
+        MENU_ITEM_SUBMENU_P(_T(MSG_REPRINT), reprint_from_eeprom);
+    }else if(!printer_active() && enableReprint && saved_printing_type == PowerPanic::PRINT_TYPE_USB) 
```
Remove white space behind `::PRINT_TYPE_USB)`

3d-gussner reviewed (changes requested):
The PR is working, thanks. Here some changes I have made:
- Rebased to MK3 branch
- Removed white spaces
- Updated Firmware.pot file with ONLY new text (using `update-pot.py `also updates the code lines which isn't useful for a PR. We use it in a pre-release PR to avoid PR conflicts.)
- Looked for existing translations in Prusa-Firmware-Buddy firmware
- Used Deepl and google translate for community languages  

Here the changes for this PR [reprint_pr_changes_diff2.txt](https://github.com/prusa3d/Prusa-Firmware/files/13267653/reprint_pr_changes_diff2.txt)

This is the diff to upstream/MK3 branch [reprint_pr_changes_diff.txt](https://github.com/prusa3d/Prusa-Firmware/files/13267631/reprint_pr_changes_diff.txt)

3d-gussner commented:
To test Host/USB `// action:reprint` I installed in OctoPrint the https://plugins.octoprint.org/plugins/actioncommands/ and configured a simple gcode action to send `M117 Reprint`
![image](https://github.com/prusa3d/Prusa-Firmware/assets/25530011/9c9d02f7-34c1-43bb-a72a-5dcb28f7694d)

jfestrada committed:
Apply the reprint_pr_changes_diff2.txt provided by @3d-gussner

3d-gussner commented:
Tested SD reprint:

1. Select SD card print 
2. When finished, choose Reprint from LCD menu starts same SD card file :heavy_check_mark: 
3. After a reset or power up with same SD card inserted the `Reprint` menu starts the last SD card file.

Test USB print:
1. Connect with host and start USB/host print
2. When finished, choose Reprint from LCD menu doesn't send via serial `// action:reprint` :exclamation:
   1. It starts the last SD card print.
3. After a reset or power up with same SD card the `Reprint` menu starts the last SD card print. :exclamation:

jfestrada commented:
> Tested SD reprint:
> 
> 1. Select SD card print
> 2. When finished, choose Reprint from LCD menu starts same SD card file ✔️
> 3. After a reset or power up with same SD card inserted the `Reprint` menu starts the last SD card file.
> 
> Test USB print:
> 
> 1. Connect with host and start USB/host print
> 2. When finished, choose Reprint from LCD menu doesn't send via serial `// action:reprint` ❗
>    
>    1. It starts the last SD card print.
> 3. After a reset or power up with same SD card the `Reprint` menu starts the last SD card print. ❗

Let me check, this is something related with the initialization of the _enableReprint_ variable, I mean, before to include the USB action, the variable was initialized to true when the card reported the EOF. Now must be set to true when the cmdqueue buffer is empty, but maybe I shall add the check that the printer was printing previously, because after a reboot the _enableReprint_ must be false.

gudnimg committed:
PFW-1523 Update M79 description

gudnimg committed:
PFW-1523 Trim trailing whitespace in util.h

leptun commented on lang/po/Firmware_ro.po:
```diff
@@ -2578,6 +2578,11 @@ msgstr "Sensibilitate"
+#. MSG_REPRINT c=18
+#: ../../Firmware/messages.cpp:190 ../../Firmware/ultralcd.cpp:5191
+msgid "Reprint"
+msgstr "Retipărire"
```
`Repetă print` to be consistent with `Print from SD`

3d-gussner replied:
@jfestrada Please change it

ingbrzy commented on lang/po/Firmware_sk.po:
```diff
@@ -2560,6 +2560,11 @@ msgstr "Citlivosť"
 
+#. MSG_REPRINT c=18
+#: ../../Firmware/messages.cpp:190 ../../Firmware/ultralcd.cpp:5191
+msgid "Reprint"
+msgstr "Predtlač"
```
Should be "Vytlačiť znova"

3d-gussner replied:
@jfestrada please change it as well

jfestrada committed:
Updates to translations of "Reprint" to Slovak and Romanian languages

jfestrada committed:
Fixed Reprint from USB and from SD, some debug is needed

jfestrada commented:
> > Tested SD reprint:
> > 
> > 1. Select SD card print
> > 2. When finished, choose Reprint from LCD menu starts same SD card file ✔️
> > 3. After a reset or power up with same SD card inserted the `Reprint` menu starts the last SD card file.
> > 
> > Test USB print:
> > 
> > 1. Connect with host and start USB/host print
> > 2. When finished, choose Reprint from LCD menu doesn't send via serial `// action:reprint` ❗
> >    
> >    1. It starts the last SD card print.
> > 3. After a reset or power up with same SD card the `Reprint` menu starts the last SD card print. ❗
> 
> Let me check, this is something related with the initialization of the _enableReprint_ variable, I mean, before to include the USB action, the variable was initialized to true when the card reported the EOF. Now must be set to true when the cmdqueue buffer is empty, but maybe I shall add the check that the printer was printing previously, because after a reboot the _enableReprint_ must be false.

Tested SD reprint:

1. Select SD card print ✔️
2. When finished, choose Reprint from LCD menu starts same SD card file ✔️
3. After a reset or power up with same SD card inserted the `Reprint` menu starts the last SD card file. ✔️

Test USB print:

1. Connect with host and start USB/host print
2. When finished, choose Reprint from LCD menu doesn't send via serial `// action:reprint` ✔️

Tomorrow I will remove the extern bool enableReprint and just use a enableReprint, also I will debug a little bit more, now the enableReprint is set to true on the function _lcdui_print_percent_done_ also I have added an enableReprintUSB in order to control when must be sent an _action:reprint_ to the host.

3d-gussner committed:
Update doxygen for M709

jfestrada committed:
Fix: Reprint menu does not appear during the heating

AttilaSVK commented:
Hi! The Hungarian translation is correct.

jfestrada ***@***.***> ezt írta (időpont: 2023. nov. 7., K,
16:48):

> This PR is a the same than #3071
> <https://github.com/prusa3d/Prusa-Firmware/pull/3071> but using the
> current repository of my fork of your firmware
>
> To support this functionality I have added a *extern bool variable* which
> can be read/set from cmdqueue.cpp and *ultralcd.cpp* so when the card
> reader reaches the EOF the variable isPrintFinished will be set to true and
> the main menu will show the *REPRINT* option.
> How do I know which file we need to reprint? All the information needed to
> perform the reprint was stored on the EEPROM in case of a power panic and
> the fw recover this information in order to start the print from SD again,
> so using that information the function reprint_from_eeprom() has been
> implemented.
>
> Translations:
>
>    - 🇨🇿 same as buddy firmware
>    <https://github.com/prusa3d/Prusa-Firmware-Buddy/blob/bfb0ffc745ee6546e7efdba618d0e7c0f4c909cd/src/lang/po/cs/Prusa-Firmware-Buddy_cs.po#L3647-L3648>
>    - 🇩🇪 same as buddy firmware
>    <https://github.com/prusa3d/Prusa-Firmware-Buddy/blob/bfb0ffc745ee6546e7efdba618d0e7c0f4c909cd/src/lang/po/de/Prusa-Firmware-Buddy_de.po#L3737-L3738>
>    - 🇪🇸 same as buddy firmware
>    <https://github.com/prusa3d/Prusa-Firmware-Buddy/blob/bfb0ffc745ee6546e7efdba618d0e7c0f4c909cd/src/lang/po/es/Prusa-Firmware-Buddy_es.po#L3710-L3711>
>    - 🇫🇷 same as buddy firmware
>    <https://github.com/prusa3d/Prusa-Firmware-Buddy/blob/bfb0ffc745ee6546e7efdba618d0e7c0f4c909cd/src/lang/po/fr/Prusa-Firmware-Buddy_fr.po#L3676-L3677>
>    - 🇮🇹 same as buddy firmware
>    <https://github.com/prusa3d/Prusa-Firmware-Buddy/blob/bfb0ffc745ee6546e7efdba618d0e7c0f4c909cd/src/lang/po/it/Prusa-Firmware-Buddy_it.po#L3698-L3699>
>    - 🇵🇱 same as buddy firmware
>    <https://github.com/prusa3d/Prusa-Firmware-Buddy/blob/bfb0ffc745ee6546e7efdba618d0e7c0f4c909cd/src/lang/po/pl/Prusa-Firmware-Buddy_pl.po#L3679-L3680>
>    - 🇳🇱 by @3d-gussner <https://github.com/3d-gussner> @vintagepc
>    <https://github.com/vintagepc> please check Firmware_nl.po
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-92de134d7cfe75b62e86a4e83716cd95550b11f2ddee228150a9c6faa4d6a34bR2585-R2586>
>    - 🇭🇷 @Prime1910 <https://github.com/Prime1910> please check
>    Firmware_hr.po
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-487d5469566575627c7dc47abf437c52e890de941c1fd8d4db5a2a4d6ef08d64R2577-R2578>
>    - 🇭🇺 @AttilaSVK <https://github.com/AttilaSVK> please check
>    Firmware_hu.po
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-3a7d51d07fb16c60d273b7986aa88a7892939b19a9bad4f3501a153ffb0604e9R2582-R2583>
>    - 🇳🇴 @OS-kar <https://github.com/OS-kar> and @trondkla
>    <https://github.com/trondkla> please check Firmware_no.po
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-22e5dd5d54111e837f3b40db65f2a126b09cc4157dcfa1f34b6e5dde0bf4a607R2559-R2560>
>    - 🇷🇴 @leptun <https://github.com/leptun> and @Hauzman
>    <https://github.com/Hauzman> please check Firmware_ro.po
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-6774f424618c20ccb1d99bf93cb34bab0b6ee03a55868fa7021a11ef6c28667fR2583-R2584>
>    - 🇸🇰 @ingbrzy <https://github.com/ingbrzy> and @shatter136
>    <https://github.com/shatter136> please check Firmware_sk.po
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-9584ba5c6a2b7618c16dc96fa21d692e9d5d69c393c55f37166bb05e98557fbcR2565-R2566>
>    - 🇸🇪 @Painkiller56 <https://github.com/Painkiller56> please check
>    Firmware_sv.po
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-cf46655e0192f28112da2985d28dc728949010730a0ac629dbb7049f9f53a357R2572-R2573>
>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/prusa3d/Prusa-Firmware/pull/4462
> Commit Summary
>
>    - 50069a4
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/commits/50069a451a00dbb6afc6a07c2948aa643fa942d6>
>    Reprint function integrated on 3.13v
>    - 7925241
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/commits/7925241ef0159c69449cbe9411dd8759f60450b6>
>    Fix the length of the PL translation of MSG_REPRINT
>    - 0eba0a2
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/commits/0eba0a2409d6110914ff92e94bfa8ad396245ca0>
>    Fix SV translation length of MSG_REPRINT
>    - 8c65187
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/commits/8c65187137fdbfaff7b31fb778507fe92e507de0>
>    Change to enquecommandf_P and enquecommand_P
>    - af61fdf
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/commits/af61fdfa55a855159094bb688b5162cec184cd8a>
>    Remove MYSERIAL.println code (debug purpose) and old commented code
>    - d052b5c
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/commits/d052b5cfaa7831a4f633e18641448befc725299b>
>    Remove some MYSERIAL calls
>    - 31d6c11
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/commits/31d6c11c44bccba2d7a5618e5071b7d4382b2d33>
>    - Serial message --> /// action:reprint done
>    - ead754c
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/commits/ead754c7a0c46052939af67f84f8cfac1b86483d>
>    Added support for reprint and streamed GCODE
>    - c44f62e
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/commits/c44f62e3e383cce255fd99e38f2343ceed6e6aca>
>    Execution of lang/update-pot.py and lang/update-po.py
>    - 934267e
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/commits/934267eeb1426a471023bfa616ecc78ce560f673>
>    Fix execution of update-po.py
>    - 1813220
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/commits/18132201a7f9a25b8fd689e07158560b91fd616f>
>    Apply the reprint_pr_changes_diff2.txt provided by @3d-gussner
>
> File Changes
>
> (21 files <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files>)
>
>    - *M* Firmware/cardreader.cpp
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-533294493df865ca3fdc626d3ed0ec2271c71c890cfc54f03e647d0a7417bd61>
>    (13)
>    - *M* Firmware/cardreader.h
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-3571f783d3944ccb93cbdb2af58b7aea62205801ba1efb65ffaa3ed5f4ae3344>
>    (3)
>    - *M* Firmware/cmdqueue.cpp
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-34d52ffb78f919a75ed3eccf9dad9c225c08d20574a803ea403943351752e2cc>
>    (1)
>    - *M* Firmware/messages.cpp
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-28c71343cce208b5091b4d61b812f185c49cca126015c962274f29e41ba3e539>
>    (3)
>    - *M* Firmware/messages.h
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-869453aa4bef38cd113b3c0932dfb7351519b9ea05695cc2c4d97ee357c3b970>
>    (2)
>    - *M* Firmware/ultralcd.cpp
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-5fbfdca43c192573e31d8e78d255c510fb6e213b118cf38173277ec2a2163880>
>    (63)
>    - *M* Firmware/ultralcd.h
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-8b5d4c1b51682b6c1b61983d821ec92bdc651b420812510fd5543cdd9bd61847>
>    (5)
>    - *M* lang/po/Firmware.pot
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-14fedb8dfdce511a0bdcf251bbe0fc1c53d69fcec094186519cc6f57826b4b1b>
>    (5)
>    - *M* lang/po/Firmware_cs.po
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-0aa4aafa0e88956cdefd904d3a4d90d8075a6baea4b7a7552fa25663b8550f7c>
>    (5)
>    - *M* lang/po/Firmware_de.po
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-7aa3bbee27c68325a9362a2508a9dbdffb40452a4296772d7b6075b7cafd7d43>
>    (5)
>    - *M* lang/po/Firmware_es.po
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-94fdeb509cca7d35e68df03af06931c6c67805acb0c427946922c1b5caa91549>
>    (5)
>    - *M* lang/po/Firmware_fr.po
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-e1c029266b2a50654a6638406d3cfe838eeb4502bde0aadca6cdcb4372d42a44>
>    (6)
>    - *M* lang/po/Firmware_hr.po
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-487d5469566575627c7dc47abf437c52e890de941c1fd8d4db5a2a4d6ef08d64>
>    (5)
>    - *M* lang/po/Firmware_hu.po
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-3a7d51d07fb16c60d273b7986aa88a7892939b19a9bad4f3501a153ffb0604e9>
>    (5)
>    - *M* lang/po/Firmware_it.po
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-d948679300ec3cbaa7e705150dfcaee8932cf6f047e800ca1036ff551f31e81c>
>    (5)
>    - *M* lang/po/Firmware_nl.po
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-92de134d7cfe75b62e86a4e83716cd95550b11f2ddee228150a9c6faa4d6a34b>
>    (5)
>    - *M* lang/po/Firmware_no.po
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-22e5dd5d54111e837f3b40db65f2a126b09cc4157dcfa1f34b6e5dde0bf4a607>
>    (5)
>    - *M* lang/po/Firmware_pl.po
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-67132bccb4c23fbeaf41753633298c3047956124c6a42a0a8a574e3c042a09e0>
>    (5)
>    - *M* lang/po/Firmware_ro.po
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-6774f424618c20ccb1d99bf93cb34bab0b6ee03a55868fa7021a11ef6c28667f>
>    (5)
>    - *M* lang/po/Firmware_sk.po
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-9584ba5c6a2b7618c16dc96fa21d692e9d5d69c393c55f37166bb05e98557fbc>
>    (5)
>    - *M* lang/po/Firmware_sv.po
>    <https://github.com/prusa3d/Prusa-Firmware/pull/4462/files#diff-cf46655e0192f28112da2985d28dc728949010730a0ac629dbb7049f9f53a357>
>    (5)
>
> Patch Links:
>
>    - https://github.com/prusa3d/Prusa-Firmware/pull/4462.patch
>    - https://github.com/prusa3d/Prusa-Firmware/pull/4462.diff
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/prusa3d/Prusa-Firmware/pull/4462>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIZ6Q5O6LDVE3AVWHREPFELYDJJ3NAVCNFSM6AAAAAA6KIMK5CVHI2DSMVQWIX3LMV43ASLTON2WKOZRHE2TKNJYGAYDQMA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

jfestrada committed:
Fix: There was a situation where enableRprintUSB was enabled but the new  print was using the SD the reprint was not working from the SD, now is fixed

jfestrada commented:
> > > Tested SD reprint:
> > > 
> > > 1. Select SD card print
> > > 2. When finished, choose Reprint from LCD menu starts same SD card file ✔️
> > > 3. After a reset or power up with same SD card inserted the `Reprint` menu starts the last SD card file.
> > > 
> > > Test USB print:
> > > 
> > > 1. Connect with host and start USB/host print
> > > 2. When finished, choose Reprint from LCD menu doesn't send via serial `// action:reprint` ❗
> > >    
> > >    1. It starts the last SD card print.
> > > 3. After a reset or power up with same SD card the `Reprint` menu starts the last SD card print. ❗
> > 
> > 
> > Let me check, this is something related with the initialization of the _enableReprint_ variable, I mean, before to include the USB action, the variable was initialized to true when the card reported the EOF. Now must be set to true when the cmdqueue buffer is empty, but maybe I shall add the check that the printer was printing previously, because after a reboot the _enableReprint_ must be false.
> 
> Tested SD reprint:
> 
>     1. Select SD card print ✔️
> 
>     2. When finished, choose Reprint from LCD menu starts same SD card file ✔️
> 
>     3. After a reset or power up with same SD card inserted the `Reprint` menu starts the last SD card file. ✔️
> 
> 
> Test USB print:
> 
>     1. Connect with host and start USB/host print
> 
>     2. When finished, choose Reprint from LCD menu doesn't send via serial `// action:reprint` ✔️
> 
> 
> Tomorrow I will remove the extern bool enableReprint and just use a enableReprint, also I will debug a little bit more, now the enableReprint is set to true on the function _lcdui_print_percent_done_ also I have added an enableReprintUSB in order to control when must be sent an _action:reprint_ to the host.

I have fixed the situation where it was printing from USB, and after that a new print from SD is started, then the reprint was using the one for USB instead of the SD, now that situation is fixed, just setting to false the _enableReprintUSB_ when is printing from SD.

3d-gussner commented:
@jfestrada Still have the issue that the Reprint is shown after a reset or power up. A Reprint after an USB print starts the last saved SD print.

jfestrada commented:
> @jfestrada Still have the issue that the Reprint is shown after a reset or power up. A Reprint after an USB print starts the last saved SD print.

That's doesn't make sense at all, I have tested several times with the new initialization of  the _enableReprint_ variable, are you sure that you are using the last commit from this PR? Now the "Reprint" menu is not enabled after a powercycle or a reboot.

I have made the following test:

1. Send a USB print --> Reprint appears --> restart the printer --> Reprint does not appears
2. Send a SD print --> Reprint appears --> restart the printer --> Reprint does not appears

I also tested: 

1. Send a USB print --> Send a SD print --> Select Reprint --> SD print start to print again
2. Send a SD print --> Send a USB print --> Select Reprint --> Octoprint receive the action:reprint and send the M117 Reprint and appear on the bottom part of the LCD
3. Send a USB print --> Select Reprint --> Octoprint receive the action:reprint and send the M117 Reprint and appear on the bottom part of the LCD
4. Send a SD print --> Select Reprint --> SD print start to print again

3d-gussner committed:
Fix issues
Check printer limits
Fix crash when pausing / paused

3d-gussner committed:
Updated MMU serial status message
- Shown during bootup
- Shown on LCD menue MMU [ON|OFF]
- Shown using `M709`
- Shown when `Disable`d with MMU ERROR FW UPDATE NEEDED

3d-gussner committed:
Code size improvments

3d-gussner committed:
Update doxygen M709

gudnimg committed:
Add a comment

gudnimg committed:
PFW-1523 Only overwrite name if not SD printing

gudnimg committed:
Fix #3122

If the feedrate multiplier is changed while the print is paused
update the saved feedrate multiplier in RAM

gudnimg committed:
EEPROM: Formatting cleanup

Mark unused bytes consistently

Add "Power Panic" into the Feature column for relevant bytes

Change "MMU2/s" to "MMU"

gudnimg committed:
PFW-1206 Add M75-M78

Change in memory:
Flash: +240 bytes
SRAM: 0 bytes

3d-gussner commented:
@Prime1910 @Painkiller56 Please have a look at the translations.

gudnimg committed:
Minor wording changes

gudnimg committed:
Replace unused EEPROM entry names with placeholders

_EEPROM_FREE_NRx_ make it a bit more clear this entry is free.

Commit:
M702: fix documentation link

3d-gussner commented:
@arekm Please rebase this PR as we have some other changes create a conflict.

gudnimg committed:
MMU: Remove gcode_reset_stepper_timeout

This function was removed on the 32-bit side.
On the 8-bit the function was empty so it is safe
remove it.

3d-gussner committed:
Add printer_state.cpp/.h

3d-gussner committed:
Show Set Ready LCD menu only when M79 timer is running.

3d-gussner committed:
Fix cmake build

3d-gussner committed:
Update state

gudnimg committed:
M125: Code size optimisation

Change in memory:
Flash: -66 bytes
SRAM: -12 bytes

gudnimg committed:
M125: make pause_position static

gudnimg committed:
Fix build failure

gudnimg committed:
M77: Save statistics when timer is stopped

For remote hosts, when the timer is stopped, then also save the statistics.

Slightly refactored save_statistics function by removing the parameters.
The function parameters are always the same.

Change in memory:
Flash: -40 bytes
SRAM: 0 bytes

gudnimg committed:
Show statistics menu for USB/Serial prints

Change in memory:
Flash: -8 bytes
SRAM: 0 bytes

gudnimg committed:
Don't reset pause_position on every M125

Change in memory:
Flash: -76 bytes
SRAM: 0 bytes

3d-gussner committed:
Don't pause with `S`et values
Update documentation

3d-gussner committed:
Raise `END_FILE_SECTION` from 20000 to 30720 to prevent false positive `File incomplete. Continue Anyway?` messages.

3d-gussner commented:
@jfestrada Tested the PR and had to fix a minor conflict with current MK3 branch. It is working well with SD prints and Host/USB prints as well.
Please rebase this PR for the last time :pray:

jfestrada commented:
> @jfestrada Tested the PR and had to fix a minor conflict with current MK3 branch. It is working well with SD prints and Host/USB prints as well. Please rebase this PR for the last time 🙏

I've made the rebase, resolve the conflict accepting the changes that are coming from my branch and should be ready to merge.

3d-gussner committed:
LCD menu only sends `// action:ready` or `// action:not_ready` via serial
Host has to send back state with `M72 S1|0` to keep printer and host in sync
Changed LCD menu messages to `Set ready` and `Set not ready`
Updated translations

3d-gussner committed:
Cleanup documentation

3d-gussner committed:
Add `SetPrinterState(PrinterState::Idle);` after `lcd_commands_type = LcdCommands::Idle;`

3d-gussner commented:
@jfestrada Looks like your rebase didn't work correctly. Can you also please update your base branch to the current Prusa MK3 branch as there are so many changed files shown in the PR.

3d-gussner commented:
@jfestrada I modified your PR in this diff a little bit after we will merge  #4429 please have a look
[Reprint_after_PR4429.diff.txt](https://github.com/prusa3d/Prusa-Firmware/files/13429851/Reprint_after_PR4429.diff.txt)
The Host Reprint is only shown when `M79` timer is active which means that the host is pining the printer and printer "knows" there is a host.

gudnimg committed:
Fix build failure

Another isPrintPaused was added by rebase

3d-gussner committed:
Whitespace

3d-gussner committed:
Rebase fix
Add documentation
Whitespace

ingbrzy reviewed (approved):
SK looks OK

jfestrada committed:
Revert "Rebase fix"

This reverts commit a9e5f445cfae6ea23473ea369b04f1eed65c21d5.

jfestrada committed:
Revert "Whitespace"

This reverts commit 5e6671a6138ef1c48e748633bc6819a76ff2b182.

jfestrada committed:
Revert "Add `SetPrinterState(PrinterState::Idle);` after `lcd_commands_type = LcdCommands::Idle;`"

This reverts commit 9970183677988f14ef1401fd73e84f324468316c.

jfestrada committed:
Revert "Cleanup documentation"

This reverts commit fcee741c16c5b362104e1e7e11431596ad5155c4.

jfestrada committed:
Revert "LCD menu only sends `// action:ready` or `// action:not_ready` via serial"

This reverts commit 17fad7dbdfe11b9bde6a8183c02637d5e8fca665.

jfestrada committed:
Revert "Update state"

This reverts commit 7a5154be5c370e2c6caa25f2e9c2153fd55f14db.

jfestrada committed:
Revert "Fix cmake build"

This reverts commit f356064f70acc307744a5d327328855692ab415c.

jfestrada committed:
Revert "Show Set Ready LCD menu only when M79 timer is running."

This reverts commit ec2bd93117222307d2f5d0d78b8f191ea4bd8567.

jfestrada committed:
Revert "Add printer_state.cpp/.h"

This reverts commit 6fceff28a9c5adc92e99634104917fcdaf3b1376.

jfestrada committed:
Revert "Set Ready"

This reverts commit ff670be16c0d5e78290da31ad41410ef09f0b20c.

jfestrada committed:
Revert "Fix build failure"

This reverts commit de80a73859b9c3657945b35a779ee4defb28373c.

jfestrada committed:
Revert "Show statistics menu for USB/Serial prints"

This reverts commit 55830e03732ae6f2272060f87374308f020ac182.

jfestrada committed:
Revert "M77: Save statistics when timer is stopped"

This reverts commit 67026f8a4baf780dd8f140562c6aa625305b30b6.

jfestrada committed:
Revert "Fix build failure"

This reverts commit 0c65e9a14fecff726f11e9c73ddcde40a441d3cb.

jfestrada committed:
Revert "PFW-1206 Add M75-M78"

This reverts commit 0df2130bf3d9f401f0961302dcd427dd5e44595d.

jfestrada committed:
Revert "Implement Marlin's print job timer"

This reverts commit 920fc60cbdfacf291ca3f210e82cd39551165316.

jfestrada committed:
Revert "PFW-1523 Implement S parameter for M79"

This reverts commit fff153828e880a7dd54e8b9c2ce5a950c2847818.

jfestrada committed:
Revert "Raise `END_FILE_SECTION` from 20000 to 30720 to prevent false positive `File incomplete. Continue Anyway?` messages."

This reverts commit 22cc4bfe4354f0625f97a2f0cba6d8f426d003a6.

jfestrada committed:
Revert "Don't pause with `S`et values"

This reverts commit 6e94ca08b598f0507d7f11bf22b6060a7547fb2d.

jfestrada committed:
Revert "Don't reset pause_position on every M125"

This reverts commit a05a66946a112a1ae4e541fdefe7d914b9828765.

jfestrada committed:
Revert "M125: make pause_position static"

This reverts commit 252c5167ed2ff4dcdad44b73db82118d3e08e295.

jfestrada committed:
Revert "M125: Code size optimisation"

This reverts commit d3d1a72cd1ce962b876119da5b5455dd7929bdbd.

jfestrada committed:
Revert "Code size improvments"

This reverts commit ac0bdb45041e5572897a0780a127cba99052093f.

jfestrada committed:
Revert "Fix issues"

This reverts commit d9428e0e5acd248f2169bd0eaa6d76dfeb1dac09.

jfestrada committed:
Revert "Fix float"

This reverts commit 57228438fa073e99df4ca7f203380c94ecbe301e.

jfestrada committed:
Revert "Add M125"

This reverts commit 1b9704fab98650840ba636cef7fb005199c0cd2c.

jfestrada committed:
Revert "optimisation: use disable_heater() in more places"

This reverts commit e14984de539710f5983eae3217bde59fa5f78fd7.

jfestrada committed:
Revert "Basic Printer name in EEPROM"

This reverts commit 6ae943206681374706369a4c349fe66d5b7db5cd.

jfestrada committed:
Revert "MMU: Remove gcode_reset_stepper_timeout"

This reverts commit fde0ab85c14d458c3cf69207a97cadee7374d573.

jfestrada committed:
Revert "Replace unused EEPROM entry names with placeholders"

This reverts commit 07556035c4d421d99dacc3fe39ecf1daf470db7c.

jfestrada committed:
Revert "Minor wording changes"

This reverts commit d49bcc2a9cba2c5976b5142a6e977812af61f124.

jfestrada committed:
Revert "EEPROM: Formatting cleanup"

This reverts commit 940b14a71f6e4158f94ca02b342a2030e7cfe653.

jfestrada committed:
Revert "Add a comment"

This reverts commit 4f8838afb8e0ac2c3542fa7d1e71fd75897afcac.

jfestrada committed:
Revert "Fix filament change behavior when print is paused"

This reverts commit e0bc60475c6ae4ab1a6671d1066f87a65ebcb5f7.

jfestrada committed:
Revert "optimisation: simplify two LCD messages"

This reverts commit 23ee18ce619d67395042fe418b0d5ac6f0f5d509.

jfestrada committed:
Revert "M702: fix documentation link"

This reverts commit 529b61b1b34c3006968afc6fb438e916a1c6442f.

jfestrada committed:
Revert "Fix #3122"

This reverts commit 74aa991193c5dfb29cbde5145b7de1b861c71b53.

jfestrada committed:
Revert "PFW-1523 Only overwrite name if not SD printing"

This reverts commit d3f055731586ddd8dd4d4d921b6fcb0ed49a27c2.

jfestrada committed:
Revert "PFW-1523 Trim trailing whitespace in util.h"

This reverts commit d2809b024f27a1dd1d4bca5c34c45016e6b87a6b.

jfestrada committed:
Revert "PFW-1523 Update M79 description"

This reverts commit d3f2de29dac19af8825e31f491d863b443ed3627.

jfestrada committed:
Revert "PFW-1523 Force inline unquoted_string constructor"

This reverts commit b7551259e31e1aecf25ff4f497521ed7b6830c70.

jfestrada committed:
Revert "PFW-1523 Implement S parameter for M79"

This reverts commit 650480d1fdffe4b91b122d2ee9818e72b710947d.

jfestrada committed:
Revert "PFW-1523 Refactor parsing quoted strings"

This reverts commit 59808097e7f6c3d321c2c1ffcea11764ddd30603.

jfestrada committed:
Revert "Update doxygen M709"

This reverts commit c32218d58e87f8f693febf68d9b2e687075f49f3.

jfestrada committed:
Revert "Updated MMU serial status message"

This reverts commit 6ca31fa5d149c61c3b04dcbe7b1f4fcc615a48d1.

jfestrada committed:
Revert "Update doxygen for M709"

This reverts commit 6e30ed3155f37e1b84209f160de4e13b62edaebf.

jfestrada committed:
Revert "Update requested changes"

This reverts commit 4e3cd06992a4e603093b0038bc3c660283d0579f.

jfestrada committed:
Revert "Add en- and disable MMU to M709"

This reverts commit 38e037848a682d4dd62c5ea222f6e1f59ae019c3.

jfestrada committed:
Revert "power panic: support any valid DOS 8.3 extension"

This reverts commit 51017ee746ac5dc7bf3012af056976aa7739330d.

jfestrada committed:
Revert "power panic: read whole directory name block from EEPROM"

This reverts commit b868dd35304d42063643681ddf1caa10aec11bf6.

jfestrada committed:
Revert "power panic: simplify recovering SD filename"

This reverts commit 6cba5fd1c17c6fb1ce9e9c47fdf2bfa880f83e54.

jfestrada committed:
Revert "Refactor menu_action_sdfile"

This reverts commit ace9012ca915cdba07e3db2116fec385e41611f8.

jfestrada committed:
Revert "power panic: read EEPROM_FILENAME until \0 is seen"

This reverts commit 7ea2c4f476ea66959e23038357ba50901119da9b.

jfestrada committed:
Revert "power panic: only send M23 for SD prints"

This reverts commit 43c0379c0ccc183e612eb0e5a66faff63fb3b0ff.

jfestrada committed:
Revert "MMU: Set PAUSED_FOR_USER state on MMU error screen"

This reverts commit 99ff9f71c054b86156dc8c69deb93c6c72dbcfd1.

jfestrada committed:
Revert "PFW-1523 Implement M79 timer for PrusaLink"

This reverts commit 237db89c0227b35742a99c57ad374b7a448266d5.

jfestrada committed:
Revert "Fix regression in MMU enable menu"

This reverts commit ff25f38439f97f549931042f1c22e75ea830caf2.

jfestrada committed:
Revert "TM: Optimise EEPROM read and writes"

This reverts commit 96db34acc8546018ec8cd15cec08a8e3b1e0f01a.

jfestrada committed:
Revert "MBL: optimise EEPROM read/writes"

This reverts commit 451f60453d4505e19f55696c3da4a778b1de4adf.

jfestrada committed:
Revert "mmu: remove isErrorScreenRunning()"

This reverts commit c68e26d2b9c6871b644a09a6cbfa3132b74ce699.

jfestrada committed:
Revert "mmu: remove obsolete EEPROM comments"

This reverts commit db0da612ead1292aafd3053ebef03c631d7ec118.

jfestrada committed:
Revert "mmu: remove ResetOnExit"

This reverts commit 3b3bdcc4239e0979ed8c6548183c1139a951df24.

jfestrada committed:
Revert "mmu: add ResetCommunicationTimeoutAttempts"

This reverts commit 3bcbfdb947a361a69fa4d6f4c339b7494c53d44f.

jfestrada committed:
Revert "mmu: add DisableMMUInSettings"

This reverts commit 2fab9d5cd86edf401cd93f1a745d3d9b008f2ddc.

jfestrada committed:
Revert "mmu: formatting"

This reverts commit 53c9d65e03374d9407e0deae7a277aee1d1155af.

jfestrada committed:
Revert "mmu: Add CheckErrorScreenUserInput"

This reverts commit 256ebb6ce8b599f89f486e866d8b3e405040612f.

3d-gussner commented:
Closing in favor of https://github.com/prusa3d/Prusa-Firmware/pull/4510

@jfestrada Thanks for this PR :heart_eyes:
